### PR TITLE
feat: add custom AMI for managed node groups

### PR DESCRIPTION
**Pull request by starchx**
*Issue #, if available:*
https://github.com/aws-quickstart/ssp-amazon-eks/issues/137

*Description of changes:*

Add custom AMI support for managed node groups

*Note:*

I had to replace this line in `stacks/eks-blueprint-stack.ts` to test this change. I am wondering you guys must have better way to handle this situation. Any suggestions are much appreciated.

```
        - const clusterProvider = blueprintProps.clusterProvider ?? new MngClusterProvider({ version });

        + // FIXME: add these to test custom AMI PR.
        + const userData = ec2.UserData.forLinux();
        + userData.addCommands(`/etc/eks/bootstrap.sh ${this.node.id}`);
        + const clusterProvider = blueprintProps.clusterProvider ?? new MngClusterProvider({
        +     version,
        +     customAmi: ec2.MachineImage.genericLinux({'ap-southeast-2': 'ami-0be34337b485b2609'}),
        +     userData,
        + });
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/cluster-providers/mng-cluster-provider.md [hunk 1]
```diff
@@ -72,4 +74,21 @@ const props: EC2ProviderClusterProps = {
...
+
+## Creating Clusters with custom AMI for the node group
+
+To create clusters using custom AMI for the worker nodes, set the `customAmi` to your custom image and provide your `userData` for node bootstrapping. 
+
+```typescript
+const userData = UserData.forLinux();
+userData.addCommands(`/etc/eks/bootstrap.sh ${cluster.clusterName}`);
+
+const props: EC2ProviderClusterProps = {
```
Comment (shapirov103) [inline docs/cluster-providers/mng-cluster-provider.md:87]: These are now called `MngClusterProviderProps`.

File: lib/cluster-providers/mng-cluster-provider.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import { Construct } from "@aws-cdk/core";
 import * as eks from "@aws-cdk/aws-eks";
 import { AutoScalingGroup } from "@aws-cdk/aws-autoscaling";
-import { InstanceType, IVpc, SubnetSelection, SubnetType } from "@aws-cdk/aws-ec2";
+import { IMachineImage, InstanceType, IVpc, LaunchTemplate, SubnetSelection, SubnetType, UserData } from "@aws-cdk/aws-ec2";
```
Comment (shapirov103) [inline lib/cluster-providers/mng-cluster-provider.ts:4]: Minor: please use wildcard import if the number of imports exceeds 3 (it is an internal convention atm, we will add it to lint): ` import * as ec2 from ...`

File: lib/cluster-providers/mng-cluster-provider.ts [hunk 3]
```diff
@@ -55,6 +55,16 @@ export interface MngClusterProviderProps extends eks.CommonClusterOptions {
...
 
+    /**
+     * The custom AMI for the node group. `amiType` and `amiReleaseVersion` will be ignored if this is set.
+     */
+    customAmi?: IMachineImage;
+
+    /**
+     * The userData for worker node when using custom AMI. Only applicable when customAmi is used.
+     */
+    userData?: UserData;
```
Comment (shapirov103) [inline lib/cluster-providers/mng-cluster-provider.ts:66]: This is confusing as on the surface it appears to be unrelated to the custom AMI. 
It would be easier to comprehend if you use a nested structure like 
```
customAmi?: {
    machineImage: IMachineImage,
    userData?: UserData
}
```

File: lib/cluster-providers/mng-cluster-provider.ts [hunk 4]
```diff
@@ -115,16 +125,33 @@ export class MngClusterProvider implements ClusterProvider {
...
             instanceTypes,
             minSize,
             maxSize,
             desiredSize,
-            releaseVersion
-        });
+        };
+        const nodegroupProps = this.props.customAmi ? {
+            ...commonNodegroupProps,
+            launchTemplateSpec: {
```
Comment (shapirov103) [inline lib/cluster-providers/mng-cluster-provider.ts:144]: We should do this only if custom AMI was provided by the user and omit otherwise.

**Review by shapirov103 — CHANGES_REQUESTED** [review]
Thank you for this PR, very helpful!
A few comments for you, nothing major.

As far as testing and overall quality assurance:

We need to add a blueprint with a custom AMI to the `main.ts`. You don't need to change the blueprint to test it out, cluster provider can be specified at blueprint creation time, for example:

```
/**
 * Custom AMI blueprint
 */
ssp.EksBlueprint.builder()
        .account("123567891").region('us-west-1')
        .clusterProvider(new ssp.MngClusterProvider( {
            version: KubernetesVersion.V1_20,
            customAmi : {...}
        }))
        .addons(new ssp.ArgoCDAddOn)
        .teams(new ssp.PlatformTeam({ name: 'platform' }))
        .build();
```

**Comment by shapirov103** [comment]
Added @askulkarni2 to the review.

**Review by askulkarni2 — DISMISSED** [review]
Aside from @shapirov103, lgtm! Consider adding a more non-trivial example in the https://github.com/aws-samples/ssp-eks-patterns repo.

File: docs/cluster-providers/mng-cluster-provider.md (hunk 1 above)
Comment (starchx) [reply docs/cluster-providers/mng-cluster-provider.md]: Fixed. Thanks.

File: lib/cluster-providers/mng-cluster-provider.ts (hunk 2 above)
Comment (starchx) [reply lib/cluster-providers/mng-cluster-provider.ts]: Fixed. Thanks.

File: lib/cluster-providers/mng-cluster-provider.ts (hunk 3 above)
Comment (starchx) [reply lib/cluster-providers/mng-cluster-provider.ts]: Fixed. Good suggestion.

File: lib/cluster-providers/mng-cluster-provider.ts (hunk 4 above)
Comment (starchx) [reply lib/cluster-providers/mng-cluster-provider.ts]: I couldn't figure a good solution on this, but I have tried another way of doing this using `if`, and updated the code. Please kindly review again. Thanks :)

Comment (shapirov103) [reply lib/cluster-providers/mng-cluster-provider.ts]: There is a way to make it more elegant, but the if statement will do for now.

**Comment by starchx** [comment]
> @starchx please take a look at this comment which I attached previosly: [#193 (review)](https://github.com/aws-quickstart/ssp-amazon-eks/pull/193#pullrequestreview-750753351)
> 
> We need to be able to run e2e test on this feature.

Yep, I just added. The tricky part here is, it requires a custom AMI ID for testing. Currently I coded in this line which is not ideal. Maybe it is something to inject in the end to end test:

https://github.com/aws-quickstart/ssp-amazon-eks/pull/193/files#diff-45234580e0a1071c682c27bcf626a0d122b1a1c0847a288e76cb81d51d0dde51R45

I tested the code locally and it creates two blue print as expected. For convenience, the custom AMI can just be a copy of EKS optimized AMI for that EKS cluster version.

File: bin/main.ts [hunk 5]
```diff
@@ -28,4 +30,22 @@ new BlueprintConstruct(app,
...
+const clusterName = 'customAmi';
+const userData = ec2.UserData.forLinux();
+userData.addCommands(`/etc/eks/bootstrap.sh ${clusterName}`);
+ssp.EksBlueprint.builder()
+    .account(process.env.CDK_DEFAULT_ACCOUNT)
+    .region(process.env.CDK_DEFAULT_REGION)
+    .clusterProvider(new ssp.MngClusterProvider({
+      version: KubernetesVersion.V1_20,
+      customAmi: {
+        machineImage: ec2.MachineImage.genericLinux({'ap-southeast-2': 'ami-0be34337b485b2609'}),
```
Comment (shapirov103) [inline bin/main.ts:45]: should the region of the cluster be set tp `ap-southest-2`? CI will operate in us-west-2.

**Review by shapirov103 — DISMISSED** [review]
LGTM (pending we pass e2e test)

**Comment by shapirov103** [comment]
@starchx  can you check if you have an issue destroying the stack after provisioning? Both times e2e tests ran the stack failed to destroy, just trying to figure out if it is something specific to the CI account.

**Comment by starchx** [comment]
> @starchx can you check if you have an issue destroying the stack after provisioning? Both times e2e tests ran the stack failed to destroy, just trying to figure out if it is something specific to the CI account.
> 
> Also can you check the failing checks below (make build, make lint, make run-test, cdk list)

```
❯ make build
rm -rf dist && node node_modules/.bin/tsc
node node_modules/.bin/copyfiles lib/**/*.yaml dist/ -u 1
❯ make lint
node node_modules/.bin/eslint . --ext .js,.jsx,.ts,.tsx
❯ make run-test
npm test

> @shapirov/cdk-eks-blueprint@0.12.0 test /Users/xinchx/dev/aws/ssp-amazon-eks
> jest --verbose

 PASS  test/stacks.test.ts (7.135 s)
  ✓ Usage tracking created (301 ms)
  ✓ Blueprint builder creates correct stack (2851 ms)
  ✓ Pipeline Builder Creates correct pipeline (579 ms)
  ✓ Nested stack add-on creates correct nested stack (551 ms)

  console.log
    SSP tracking (qs-1s1r465hk)

      at Object.<anonymous> (test/stacks.test.ts:11:13)

  console.log
    My awesome description SSP tracking (qs-1s1r465hk)

      at Object.<anonymous> (test/stacks.test.ts:19:13)

  console.log
    SSP tracking (qs-1s1r465k6)

      at Object.<anonymous> (test/stacks.test.ts:90:13)

Test Suites: 1 passed, 1 total
Tests:       4 passed, 4 total
Snapshots:   0 total
Time:        7.967 s, estimated 16 s
Ran all test suites.
❯ npx cdk list
customAmi
eks-blueprint-vpc
blueprint-construct-dev
❯ cdk list
customAmi
eks-blueprint-vpc
blueprint-construct-dev
```

**Comment by starchx** [comment]
@shapirov103 , I have found the root cause of stack deletion failure, due to the `argocd` namespace stuck in terminating status. Will share more details with you in slack. It is not related to the change in this PR.

**Comment by shapirov103** [comment]
@starchx as I mentioned in slack: 
```
It is a known issue, we cover it in the argocd docs: https://github.com/aws-quickstart/ssp-amazon-eks/blob/main/docs/addons/argo-cd.md#known-issues . Unclear, why it would only manifest when your stack is in the picture. I ran e2e tests for all PRs that were merged last week with no error, so it did not occur to me that it could be the source of the problem.
Anyway path forward: please update examples/blueprint-construct/index.ts to deploy argocd add-on without workload bootstrapping. Also we need to create an issue to solve this problem for tests. Once you create the issue, please reference it is in the PR and I will rerun the test and merge it.
```

Also can you address the failing check above (failure from github action)? 

```
Run make list
npx cdk list
Unable to find AMI in AMI map: no AMI specified for region 'us-east-1'
Subprocess exited with error 1
make: *** [Makefile:24: list] Error 1
Error: Process completed with exit code 2.
```

**Comment by starchx** [comment]
Created an issue to track the stack deletion error https://github.com/aws-quickstart/ssp-amazon-eks/issues/212

**Comment by starchx** [comment]
@shapirov103 , good point on the `github action not suppose to have credential`. Based on the CI error message, I want to try this first, by adding image ID for `us-east-1` to see if this helps.
```
      machineImage: ec2.MachineImage.genericLinux({
        'us-east-1': 'ami-0b297a512e2852b89',
        'us-west-2': 'ami-06a8c459c01f55c7b',
      }),
```
Would you mind kicking off another CI for me please?
